hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/onset/OnsetChecker.java

```diff
@@ -0,0 +1,230 @@
+    public void checkSignedTransaction(@NonNull final SignedTransaction tx)
+            throws PreCheckException {
+        requireNonNull(tx);
+
+        if (MiscUtils.hasUnknownFields(tx)) {
+            throw new PreCheckException(TRANSACTION_HAS_UNKNOWN_FIELDS);
+        }
+    }
+
+    /**
+     * Validates a {@link TransactionBody}
+     *
+     * @param txBody the {@code TransactionBody} to check
+     * @throws PreCheckException if validation fails
+     * @throws NullPointerException if any of the parameters is {@code null}
+     */
+    public void checkTransactionBody(@NonNull final TransactionBody txBody)
+            throws PreCheckException {
+        requireNonNull(txBody);
+
+        if (MiscUtils.hasUnknownFields(txBody)) {
+            throw new PreCheckException(TRANSACTION_HAS_UNKNOWN_FIELDS);
+        }
+
+        if (!txBody.hasTransactionID()) {
+            throw new PreCheckException(INVALID_TRANSACTION_ID);
+        }
+
+        var txnId = txBody.getTransactionID();
+        if (recordCache.isReceiptPresent(txnId)) {
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
+
+        if (!isPlausibleTxnFee(txBody.getTransactionFee())) {
+            throw new PreCheckException(INSUFFICIENT_TX_FEE);
+        }
+
+        if (!isPlausibleAccount(txnId.getAccountID())) {
+            throw new PreCheckException(PAYER_ACCOUNT_NOT_FOUND);
+        }
+
+        checkMemo(txBody.getMemo());
+
+        checkTimebox(txnId.getTransactionValidStart(), txBody.getTransactionValidDuration());
+    }
+
+    private static boolean isPlausibleTxnFee(final long transactionFee) {
+        return transactionFee >= 0;
+    }
+
+    private static boolean isPlausibleAccount(final AccountID accountID) {
+        return accountID.getAccountNum() > 0
+                && accountID.getRealmNum() >= 0
+                && accountID.getShardNum() >= 0;
+    }
+
+    private void checkMemo(final String memo) throws PreCheckException {
+        final var buffer = memo == null ? new byte[0] : memo.getBytes(StandardCharsets.UTF_8);
+        if (buffer.length > dynamicProperties.maxMemoUtf8Bytes()) {
+            throw new PreCheckException(MEMO_TOO_LONG);
+        }
+        for (final byte b : buffer) {
+            if (b == 0) {
+                throw new PreCheckException(INVALID_ZERO_BYTE_IN_STRING);
+            }
+        }
+    }
+
+    private void checkTimebox(final Timestamp start, final Duration duration)
+            throws PreCheckException {
+        final var validForSecs = duration.getSeconds();
+        if (validForSecs < dynamicProperties.minTxnDuration()
+                || validForSecs > dynamicProperties.maxTxnDuration()) {
+            throw new PreCheckException(INVALID_TRANSACTION_DURATION);
+        }
+
+        final var validStart = safeguardedInstant(start);
+        final var validDuration = safeguardedDuration(validForSecs, validStart);
+        final var consensusTime = Instant.now(Clock.systemUTC());
```

Comment:
This class will be easier to test if we use a time provider here:

```
// production instantiation:
new OnsetChecker(() -> Instant.now(Clock.systemUTC())
// test instantiation:
new OnsetChecker(() -> Instant.ofEpochSecond(1_234_567L))

// production call:
final var consensusTime = timeProvider.get();
```
or something similar